sycl/source/detail/context_impl.cpp

```diff
@@ -394,26 +394,29 @@ void context_impl::DeviceGlobalInitializer::ClearEvents(const plugin &Plugin) {
   MDeviceGlobalInitEvents.clear();
 }
 
-std::optional<RT::PiProgram> context_impl::getProgramForDeviceGlobal(
-    const device &Device, DeviceGlobalMapEntry *DeviceGlobalEntry) {
+std::optional<RT::PiProgram>
+context_impl::getProgramForDevImgs(const device &Device,
+                                   std::set<std::uintptr_t> &ImgIdentifiers) {
+
   KernelProgramCache::ProgramWithBuildStateT *BuildRes = nullptr;
   {
     auto LockedCache = MKernelProgramCache.acquireCachedPrograms();
     auto &KeyMap = LockedCache.get().KeyMap;
     auto &Cache = LockedCache.get().Cache;
     RT::PiDevice &DevHandle = getSyclObjImpl(Device)->getHandleRef();
-    for (std::uintptr_t ImageIDs : DeviceGlobalEntry->MImageIdentifiers) {
+    for (std::uintptr_t ImageIDs : ImgIdentifiers) {
       auto OuterKey = std::make_pair(ImageIDs, DevHandle);
       size_t NProgs = KeyMap.count(OuterKey);
       if (NProgs == 0)
         continue;
       // If the cache has multiple programs for the identifiers or if we have
-      // already found a program in the cache with the device_global, we cannot
-      // proceed.
+      // already found a program in the cache with the device_global or host
+      // pipe we cannot proceed.
       if (NProgs > 1 || (BuildRes && NProgs == 1))
         throw sycl::exception(
             make_error_code(errc::invalid),
-            "More than one image exists with the device_global.");
+            "More than one image exists with the device_global or host pipe.");
```

Comment:
Instead of being vague, could we maybe pass a string with the object type and concat them here? I.e. something like

```suggestion
            "More than one image exists with the " + ContainedObjectTypeName + ".");
```